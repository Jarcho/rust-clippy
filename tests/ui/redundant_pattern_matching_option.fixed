#![warn(clippy::all)]
#![warn(clippy::redundant_pattern_matching)]
#![allow(
    unused_must_use,
    clippy::needless_bool,
    clippy::needless_if,
    clippy::match_like_matches_macro,
    clippy::equatable_if_let,
    clippy::if_same_then_else
)]
#![feature(let_chains, if_let_guard)]

fn issue_11174<T>(boolean: bool, maybe_some: Option<T>) -> bool {
    maybe_some.is_none() && (!boolean)
    //~^ redundant_pattern_matching
}

fn issue_11174_edge_cases<T>(boolean: bool, boolean2: bool, maybe_some: Option<T>) {
    //~v redundant_pattern_matching
    let _ = maybe_some.is_none() && (boolean || boolean2); // guard needs parentheses
    let _ = match maybe_some {
        // can't use `matches!` here
        // because `expr` metavars in macros don't allow let exprs
        None if let Some(x) = Some(0)
            && x > 5 =>
        {
            true
        },
        _ => false,
    };
}

fn main() {
    if None::<()>.is_none() {} //~ redundant_pattern_matching

    if Some(42).is_some() {} //~ redundant_pattern_matching

    //~v redundant_pattern_matching
    if Some(42).is_some() {
        foo();
    } else {
        bar();
    }

    while Some(42).is_some() {} //~ redundant_pattern_matching

    while Some(42).is_none() {} //~ redundant_pattern_matching

    while None::<()>.is_none() {} //~ redundant_pattern_matching

    let mut v = vec![1, 2, 3];
    //~v redundant_pattern_matching
    while v.pop().is_some() {
        foo();
    }

    if None::<i32>.is_none() {}

    if Some(42).is_some() {}

    //~v redundant_pattern_matching
    Some(42).is_some();

    //~v redundant_pattern_matching
    None::<()>.is_none();

    //~v redundant_pattern_matching
    let _ = None::<()>.is_none();

    let opt = Some(false);
    let _ = if opt.is_some() { true } else { false };
    //~^ redundant_pattern_matching

    issue6067();
    issue10726();
    issue10803();

    //~v redundant_pattern_matching
    let _ = if gen_opt().is_some() {
        1
    //~v redundant_pattern_matching
    } else if gen_opt().is_none() {
        2
    } else {
        3
    };

    if gen_opt().is_some() {} //~ redundant_pattern_matching
}

fn gen_opt() -> Option<()> {
    None
}

fn foo() {}

fn bar() {}

// Methods that are unstable const should not be suggested within a const context, see issue #5697.
// However, in Rust 1.48.0 the methods `is_some` and `is_none` of `Option` were stabilized as const,
// so the following should be linted.
const fn issue6067() {
    if Some(42).is_some() {} //~ redundant_pattern_matching

    if None::<()>.is_none() {} //~ redundant_pattern_matching

    while Some(42).is_some() {} //~ redundant_pattern_matching

    while None::<()>.is_none() {} //~ redundant_pattern_matching

    //~v redundant_pattern_matching
    Some(42).is_some();

    //~v redundant_pattern_matching
    None::<()>.is_none();
}

#[allow(clippy::deref_addrof, dead_code, clippy::needless_borrow)]
fn issue7921() {
    if (&None::<()>).is_none() {} //~ redundant_pattern_matching
    if (&None::<()>).is_none() {} //~ redundant_pattern_matching
}

fn issue10726() {
    let x = Some(42);

    //~v redundant_pattern_matching
    x.is_some();

    //~v redundant_pattern_matching
    x.is_none();

    //~v redundant_pattern_matching
    x.is_none();

    //~v redundant_pattern_matching
    x.is_some();

    // Don't lint
    match x {
        Some(21) => true,
        _ => false,
    };
}

fn issue10803() {
    let x = Some(42);

    let _ = x.is_some(); //~ redundant_pattern_matching

    let _ = x.is_none(); //~ redundant_pattern_matching

    // Don't lint
    let _ = matches!(x, Some(16));
}
