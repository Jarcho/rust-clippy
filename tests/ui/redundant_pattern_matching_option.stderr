error: redundant pattern matching, consider using `is_none()`
  --> $DIR/redundant_pattern_matching_option.rs:14:5
   |
LL |     matches!(maybe_some, None if !boolean)
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `maybe_some.is_none() && (!boolean)`
   |
   = note: `-D clippy::redundant-pattern-matching` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(clippy::redundant_pattern_matching)]`

error: redundant pattern matching, consider using `is_none()`
  --> $DIR/redundant_pattern_matching_option.rs:20:13
   |
LL |     let _ = matches!(maybe_some, None if boolean || boolean2); // guard needs parentheses
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `maybe_some.is_none() && (boolean || boolean2)`

error: redundant pattern matching, consider using `is_none()`
  --> $DIR/redundant_pattern_matching_option.rs:34:12
   |
LL |     if let None = None::<()> {}
   |     -------^^^^------------- help: try: `if None::<()>.is_none()`

error: redundant pattern matching, consider using `is_some()`
  --> $DIR/redundant_pattern_matching_option.rs:36:12
   |
LL |     if let Some(_) = Some(42) {}
   |     -------^^^^^^^----------- help: try: `if Some(42).is_some()`

error: redundant pattern matching, consider using `is_some()`
  --> $DIR/redundant_pattern_matching_option.rs:39:12
   |
LL |     if let Some(_) = Some(42) {
   |     -------^^^^^^^----------- help: try: `if Some(42).is_some()`

error: redundant pattern matching, consider using `is_some()`
  --> $DIR/redundant_pattern_matching_option.rs:45:15
   |
LL |     while let Some(_) = Some(42) {}
   |     ----------^^^^^^^----------- help: try: `while Some(42).is_some()`

error: redundant pattern matching, consider using `is_none()`
  --> $DIR/redundant_pattern_matching_option.rs:47:15
   |
LL |     while let None = Some(42) {}
   |     ----------^^^^----------- help: try: `while Some(42).is_none()`

error: redundant pattern matching, consider using `is_none()`
  --> $DIR/redundant_pattern_matching_option.rs:49:15
   |
LL |     while let None = None::<()> {}
   |     ----------^^^^------------- help: try: `while None::<()>.is_none()`

error: redundant pattern matching, consider using `is_some()`
  --> $DIR/redundant_pattern_matching_option.rs:53:15
   |
LL |     while let Some(_) = v.pop() {
   |     ----------^^^^^^^---------- help: try: `while v.pop().is_some()`

error: redundant pattern matching, consider using `is_some()`
  --> $DIR/redundant_pattern_matching_option.rs:62:5
   |
LL | /     match Some(42) {
LL | |         Some(_) => true,
LL | |         None => false,
LL | |     };
   | |_____^ help: try: `Some(42).is_some()`

error: redundant pattern matching, consider using `is_none()`
  --> $DIR/redundant_pattern_matching_option.rs:68:5
   |
LL | /     match None::<()> {
LL | |         Some(_) => false,
LL | |         None => true,
LL | |     };
   | |_____^ help: try: `None::<()>.is_none()`

error: redundant pattern matching, consider using `is_none()`
  --> $DIR/redundant_pattern_matching_option.rs:74:13
   |
LL |       let _ = match None::<()> {
   |  _____________^
LL | |         Some(_) => false,
LL | |         None => true,
LL | |     };
   | |_____^ help: try: `None::<()>.is_none()`

error: redundant pattern matching, consider using `is_some()`
  --> $DIR/redundant_pattern_matching_option.rs:80:20
   |
LL |     let _ = if let Some(_) = opt { true } else { false };
   |             -------^^^^^^^------ help: try: `if opt.is_some()`

error: redundant pattern matching, consider using `is_some()`
  --> $DIR/redundant_pattern_matching_option.rs:88:20
   |
LL |     let _ = if let Some(_) = gen_opt() {
   |             -------^^^^^^^------------ help: try: `if gen_opt().is_some()`

error: redundant pattern matching, consider using `is_none()`
  --> $DIR/redundant_pattern_matching_option.rs:91:19
   |
LL |     } else if let None = gen_opt() {
   |            -------^^^^------------ help: try: `if gen_opt().is_none()`

error: redundant pattern matching, consider using `is_some()`
  --> $DIR/redundant_pattern_matching_option.rs:97:12
   |
LL |     if let Some(..) = gen_opt() {}
   |     -------^^^^^^^^------------ help: try: `if gen_opt().is_some()`

error: redundant pattern matching, consider using `is_some()`
  --> $DIR/redundant_pattern_matching_option.rs:112:12
   |
LL |     if let Some(_) = Some(42) {}
   |     -------^^^^^^^----------- help: try: `if Some(42).is_some()`

error: redundant pattern matching, consider using `is_none()`
  --> $DIR/redundant_pattern_matching_option.rs:114:12
   |
LL |     if let None = None::<()> {}
   |     -------^^^^------------- help: try: `if None::<()>.is_none()`

error: redundant pattern matching, consider using `is_some()`
  --> $DIR/redundant_pattern_matching_option.rs:116:15
   |
LL |     while let Some(_) = Some(42) {}
   |     ----------^^^^^^^----------- help: try: `while Some(42).is_some()`

error: redundant pattern matching, consider using `is_none()`
  --> $DIR/redundant_pattern_matching_option.rs:118:15
   |
LL |     while let None = None::<()> {}
   |     ----------^^^^------------- help: try: `while None::<()>.is_none()`

error: redundant pattern matching, consider using `is_some()`
  --> $DIR/redundant_pattern_matching_option.rs:121:5
   |
LL | /     match Some(42) {
LL | |         Some(_) => true,
LL | |         None => false,
LL | |     };
   | |_____^ help: try: `Some(42).is_some()`

error: redundant pattern matching, consider using `is_none()`
  --> $DIR/redundant_pattern_matching_option.rs:127:5
   |
LL | /     match None::<()> {
LL | |         Some(_) => false,
LL | |         None => true,
LL | |     };
   | |_____^ help: try: `None::<()>.is_none()`

error: redundant pattern matching, consider using `is_none()`
  --> $DIR/redundant_pattern_matching_option.rs:135:12
   |
LL |     if let None = *(&None::<()>) {}
   |     -------^^^^----------------- help: try: `if (&None::<()>).is_none()`

error: redundant pattern matching, consider using `is_none()`
  --> $DIR/redundant_pattern_matching_option.rs:136:12
   |
LL |     if let None = *&None::<()> {}
   |     -------^^^^--------------- help: try: `if (&None::<()>).is_none()`

error: redundant pattern matching, consider using `is_some()`
  --> $DIR/redundant_pattern_matching_option.rs:143:5
   |
LL | /     match x {
LL | |         Some(_) => true,
LL | |         _ => false,
LL | |     };
   | |_____^ help: try: `x.is_some()`

error: redundant pattern matching, consider using `is_none()`
  --> $DIR/redundant_pattern_matching_option.rs:149:5
   |
LL | /     match x {
LL | |         None => true,
LL | |         _ => false,
LL | |     };
   | |_____^ help: try: `x.is_none()`

error: redundant pattern matching, consider using `is_none()`
  --> $DIR/redundant_pattern_matching_option.rs:155:5
   |
LL | /     match x {
LL | |         Some(_) => false,
LL | |         _ => true,
LL | |     };
   | |_____^ help: try: `x.is_none()`

error: redundant pattern matching, consider using `is_some()`
  --> $DIR/redundant_pattern_matching_option.rs:161:5
   |
LL | /     match x {
LL | |         None => false,
LL | |         _ => true,
LL | |     };
   | |_____^ help: try: `x.is_some()`

error: redundant pattern matching, consider using `is_some()`
  --> $DIR/redundant_pattern_matching_option.rs:176:13
   |
LL |     let _ = matches!(x, Some(_));
   |             ^^^^^^^^^^^^^^^^^^^^ help: try: `x.is_some()`

error: redundant pattern matching, consider using `is_none()`
  --> $DIR/redundant_pattern_matching_option.rs:178:13
   |
LL |     let _ = matches!(x, None);
   |             ^^^^^^^^^^^^^^^^^ help: try: `x.is_none()`

error: aborting due to 30 previous errors

